First draft of USDC migration guide

joshuahannan opened the pull request:
Adds a first draft of the USDC migration guide for users and developers.
It is probably missing some cases, but it is a good enough place to start. I'll be adding more to it in the coming days but it would be good to get something up for it for now so people can get started.

SeanRobb commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+
+# USDC in Cadence 1.0
+
+Effective September 3, 2024, at 3 a.m. ET ahead of the Crescendo network upgrade,
```
Can we update this paragraph to fit more this language?


> As part of the upcoming Flow Crescendo network upgrade, all ERC-20 compatible tokens on Flow can be used in Cadence smart contracts without special handling and all Cadence fungible tokens on Flow will be ERC-20 compatible. This will ensure that all Flow applications—whether they are built with Solidity, Cadence, or some mix of both—are fully interoperable with the wider web3 ecosystem.
> 
> Effective September 3, 2024, at 3 a.m. ET ahead of the Crescendo network upgrade, Circle will cease support for USDC on Cadence, as EVM-based ERC-20 USDC will become naturally compatible across the entire Flow ecosystem

We want the context to be given for the update and change. Framing can help people understand how this is an important step to the ecosystem.

joshuahannan replied:
added ✅

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+`USDCFlow` will seamlessly retain its 1 USD value and be able to be redeemed
+with Circle by bridging back to Ethereum mainnet.
+
+This document is primarily focused on what developers who rely on `FiatToken`
+can do to migrate to `USDCFlow` with as little disruption as possible.
+A brief guide is provided in the next section for regular USDC users
+about what they can do to migrate to the new token, but those users should primarily
+follow the [Flow Blog](https://www.flow.com/authors/flow) and communication
+from the apps they use about what their options are to migrate to the new USDC on Flow.
```
```suggestion
This document focuses on what developers who rely on `FiatToken`
can do to migrate to `USDCFlow` with minimal disruption.
`USDC` users can refer to the guide below for simple migration instructions, but it's also important to 
consult the [Flow Blog](https://www.flow.com/authors/flow) and any communication
from relevant ecosystem apps for more detail about `USDCFlow` migration.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+# What can regular USDC holders do?
+
+If you are a user with USDC in your account, the recommended path is to directly 
+swap your USDC with `USDCFlow` using this transaction. This functionality
```
```suggestion
swap your `USDC` with `USDCFlow` using this transaction. This functionality
```

gregsantos replied:
which transaction is this referring to?

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+swap your USDC with `USDCFlow` using this transaction. This functionality
+is available on [Flow Port](https://port.flow.com/) with any Flow-compatible wallet.
+
+As described in the blog post linked in the previous section, USDC holders
+can also swap to FLOW on [increment.fi](https://app.increment.fi/),
+deposit to [Dapper Wallet](https://meetdapper.com/) as Dapper balance, 
+or wait until after September 3rd and manually redeem with Circle.
```
```suggestion
As described in the blog post linked in the previous section, USDC holders have a few options for migrating to the new USDCFlow token:

1. Swap USDC for FLOW on [increment.fi](https://app.increment.fi/).
2. Deposit USDC to [Dapper Wallet](https://meetdapper.com/) as Dapper balance.
3. Wait until after September 3rd and manually redeem USDC with Circle.

Regardless of the approach, `USDC` holders should follow the guidance provided on the [Flow Blog](https://www.flow.com/authors/flow) and communications from the apps they use for specific instructions on how to migrate to the new `USDCFlow` token.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+deposit to [Dapper Wallet](https://meetdapper.com/) as Dapper balance, 
+or wait until after September 3rd and manually redeem with Circle.
+
+If you are a USDC holder on Flow and you use an app that transacts in USDC in any way,
+you should reach out to the app's support channels to see if there is anything specific
+you need to do to migrate your usage of that app to the new `USDCFlow` token.
+The Flow foundation is already working with most apps that use USDC
+to help them with their migration plans to the new `USDCFlow`,
+so rest assured that there will be a path for you to retain your assets
+past the Crescendo migration, regardless of what form that takes.
```
```suggestion
If you hold `USDC` on Flow and use any apps that transact in `USDC`, refer to the app's communication and use the app's support channels to inquire about specific migration procedures for the new USDCFlow token.
Be aware that the Flow Foundation is collaborating with most apps who are using `USDC` to facilitate their migration to `USDCFlow`.
Rest assured that there will be a way for you to retain your assets after the Crescendo migration, regardless of the specific migration path.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+
+## Flow Users with Marketplace listings for USDC
+
+If you are a user with a NFT listing on a marketplace with USDC as the payment currency,
+your existing listings will not work any more after the Crescendo upgrade.
```
```suggestion
If you have NFT listings on a marketplace that accepts `USDC` as the payment currency, your existing listings will no longer work after the Crescendo upgrade.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+
+If you are a user with a NFT listing on a marketplace with USDC as the payment currency,
+your existing listings will not work any more after the Crescendo upgrade.
+If you still want your listings to work, you will need to create new listings
+with the new `USDCFlow` currency so they are still available after the Crescendo upgrade.
```
```suggestion
To ensure your listings remain active, you will need to create new listings that accept the new `USDCFlow` currency.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+your existing listings will not work any more after the Crescendo upgrade.
+If you still want your listings to work, you will need to create new listings
+with the new `USDCFlow` currency so they are still available after the Crescendo upgrade.
+This will likely require that whatever app you were using for marketplace listings
+supports the transaction that lists for `USDCFlow` instead of `FiatToken`,
+so if this isn't immediately available, please contact their support
+to ask if it will be supported.
```
```suggestion
This will likely require the marketplace app you used for your listings to support the transaction for listing with `USDCFlow` instead of `FiatToken`. If this functionality is not immediately available, please contact the app's support team to inquire about their plans to add support for `USDCFlow` listings.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+supports the transaction that lists for `USDCFlow` instead of `FiatToken`,
+so if this isn't immediately available, please contact their support
+to ask if it will be supported.
+Luckily, this is not time sensistive and can be completed
+any time after the Crescendo migration with no downside besides
+old listings not being purchasable until new ones with `USDCFlow` are created.
```
```suggestion
Fortunately, this migration process is not time-sensitive. You can complete the transition to `USDCFlow` listings at any point after the Crescendo upgrade. The only downside is that your old listings using `USDC` will not be purchasable until you create new ones with `USDCFlow`.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+This means that if your contract relies on any of that functionality,
+you'll need to do some extra work to migrate to the new token.
+
+The first thing to consider is if you are able to withdraw the USDC from these fields
```
```suggestion
The first thing to consider is if you are able to withdraw the `USDC` from these fields
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+and deposit the new `USDCFlow` into those fields.
+
+Cadence doesn't allow adding new fields to a contract directly, but you can
+somewhat get around this restriction by adding new psuedo-fields in the form
```
```suggestion
get around this restriction by adding new psuedo-fields in the form
```

Aliserag commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+a new fungible token contract called `USDCFlow` is deployed on [testnet](https://testnet.flowdiver.io/contract/A.4516677f8083d680.USDCFlow?tab=deployments)
+and mainnet (COMING SOON).
+You can see the Github repo with the contract code and transactions to use it [here](https://github.com/onflow/bridged-usdc).
+that allows anyone with a `FiatToken` balance to deposit into the `USDCFLow` contract
```
Tiny typo: Capitalize 'T' in "That' and add period after function.

sisyphusSmiling replied:
```suggestion
that allows anyone with a `FiatToken` balance to deposit into the `USDCFlow` contract
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
````diff
@@ -0,0 +1,330 @@
+}
+```
+
+You could have some similar but slightly more complex code to move
+the code from Vaults stored in a dictionary or an array.
````
```suggestion
You could have similar, but slightly more complex, code to migrate Vaults stored in a dictionary or an array.
```

Aliserag commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+# What can regular USDC holders do?
+
+If you are a user with USDC in your account, the recommended path is to directly 
+swap your USDC with `USDCFlow` using this transaction. This functionality
```
What is 'this transaction'?

Aliserag replied:
It's not possible to do anything till it's deployed on mainnet right? Otherwise I can make a small video tutorial we can add.

joshuahannan replied:
oops, meant to link to the transaction to use. I updated it.

joshuahannan replied:
Yeah, they can't do anything on mainnet until it is deployed, but that is happening very soon

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+
+## Conclusion
+
+This list is just a high-level overview of the potential ways that Cadence
+code can use `FiatToken`. If your project uses it in a different way or 
+you need more help with migrating to the new token, please reach out
+in the Flow Discord and the Flow team will be happy to assist you with your upgrade.
```
```suggestion
This overview covers the primary ways Cadence code can interact with the `FiatToken` contract. However, if your project uses `FiatToken` in a different manner or you require further assistance with migrating to the new `USDCFlow token`, please reach out in the [Flow Discord](https://discord.com/invite/J6fFnh2xx6). The Flow team will be happy to provide guidance and help you navigate the upgrade process.
```

Aliserag commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+you'll need to do some extra work to migrate to the new token.
+
+The first thing to consider is if you are able to withdraw the USDC from these fields
+in any way. Ideally, you could withdraw the USDC from these `FiatToken` fields
```
Lets change 'FiatToken' here to USDC because regular users won't understand FiatToken if they are coming just to this section to understand next steps.

joshuahannan replied:
This is the developer section of the doc, so they should all understand what `FiatToken` is

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+As described above in the user section, all `NFTStorefront` listings that
+expect `FiatToken` to be used as the form of payment will break after the Crescendo upgrade.
+
+Any app or project that is using `NFTStorefront` or a marketplace that is similar with USDC
+should just expect all `FiatToken` listings to break and should expect users to just
+create new listings that use `USDCFlow` instead. These listings that break
+will not cause any assets to be lost or compromised. It will just require a bit of work
+to re-list all the assets that were listed before.
```
```suggestion
Any app or project using `NFTStorefront` or a similar `USDC`-based marketplace should anticipate that all `FiatToken` listings will cease to function. Users will need to create new listings using `USDCFlow` instead. It's important to note that while these broken listings won't cause any assets to be lost or compromised, it will require some effort to re-list all previously listed assets.
```

gregsantos commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,330 @@
+Luckily, this is a very rare pattern and might not exist anywhere,
+so hopefully it won't be an issue.
+
+In case it is an issue for someone, please reach out to the Flow team for assistance
+and the team can likely find a way to help you resolve this.
```
```suggestion
If this situation applies to your project, please contact the Flow team for assistance. We can work with you to find a suitable solution for resolving this issue.
```

gregsantos reviewed (commented):
We might want to document the edge case for user's using Ledger which doesn't allow for arbitrary txs. I think that excludes them from using Flow Port to exchange

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+
+If you are a user with USDC in your account, the recommended path is to directly 
+swap your USDC with `USDCFlow` using this transaction. This functionality
+is available on [Flow Port](https://port.flow.com/) with any Flow-compatible wallet.
```
Is this true for ledger as well?

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+have to do some work to remove the dependency and/or migrate
+your code and state to the new `USDCFlow` token contract.
+
+You will likely have to make multiple sets of changes if you want 
+this process to go as smoothly as possible. This obviously depends
+on the project and what kind of integration with USDC you already have.
```
```suggestion
The various changes required will depend on your exact implementation and your level of integration  with the `FiatToken` contract.
```

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+this process to go as smoothly as possible. This obviously depends
+on the project and what kind of integration with USDC you already have.
+
+The first set is the pre-crescendo changes
+to allow users to pull their USDC out of your contracts, swap to the new `USDCFlow` token,
+and then re-enter your updated contract.
+Second, you'll have to update these changes to Cadence 1.0 so that they
+continue to work after Crescendo.
```
```suggestion
It's necessary to complete the following steps **before** the Crescendo milestone:

1. To allow users to pull their USDC out of your contracts, swap to the new `USDCFlow` token,
and then re-enter your updated contract.
2. Update these changes in your Cadence 1.0 contract versions and stage the updated contract code so that they continue to work after Crescendo.
```

> then re-enter your updated contract.

Not sure what's meant by "re-enter" here

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+that point to `FiatToken` vaults. These will also need to be updated
+to the new types.
+
+If your users are expecting to use the old `FiatToken` values with your contract,
+it is important for you to notify them that in order to keep using your contract,
+they need to convert their old `FiatToken` to `USDCFlow`
+via the public `USDCFlow.wrapFiatToken()` function before interacting anymore.
+It might also be useful to provide helpful error messages in functions and transactions
+that direct users to how they can swap their `FiatToken.Vault` for a `USDCFlow.Vault`.
```
```suggestion
If your users are expecting to use the old `FiatToken` values with your contract,
it is important for you to notify them that in order to keep using your contract,
they need to convert their old `FiatToken` to `USDCFlow`
via the public `USDCFlow.wrapFiatToken()` function before they can continue interacting.

It might also be useful to provide helpful error messages in functions and transactions
that direct users to how they can swap their `FiatToken.Vault` for a `USDCFlow.Vault`.
```

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+
+    // Get a reference to the recipient's Receiver
+    let receiverRef = recipient.capabilities.borrow<&{FungibleToken.Receiver}>(USDCFlow.ReceiverPublicPath)
+        ?? panic("Could not borrow receiver reference to the recipient's Vault")
```
```suggestion
        ?? panic("Could not borrow receiver reference to the recipient's USDCFlow Vault")
```

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+    // Get a reference to the recipient's Receiver
+    let receiverRef = recipient.getCapability(FiatToken.VaultReceiverPubPath)
+        .borrow<&{FungibleToken.Receiver}>()
+        ?? panic("Could not borrow receiver reference to the recipient's Vault")
```
```suggestion
        ?? panic("Could not borrow receiver reference to the recipient's USDC Vault")
```

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+
+## Stored `@FiatToken.Vault` Field
+
+If your contract stores ANY types from the `FiatToken` contract, 
+this means that you have state that will not work properly after the Crescendo upgrade.
```
I get what's being said here, but it could be unclear based on the content above - it's stated above that the FiatToken vault will pass type checking. Suggesting below based on my interpretation, but I could also be wrong 

```suggestion
If your contract stores **ANY** values defined in `FiatToken` contract, 
**you have state that will not work properly after the Crescendo upgrade.**
```

joshuahannan replied:
This seems fine to me

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
````diff
@@ -0,0 +1,336 @@
+```
+
+It is more challenging to deal with vaults that are stored in a composite type
+becuase those can live in any account. Fixing them will be more difficult because
````
```suggestion
because those can live in any account. Fixing them will be more difficult because
```

sisyphusSmiling commented on versioned_docs/version-1.0/cadence-migration-guide/usdc-migration.mdx:
```diff
@@ -0,0 +1,336 @@
+becuase those can live in any account. Fixing them will be more difficult because
+you have to work with users to change it. 
+
+Luckily, this is a very rare pattern and might not exist anywhere,
+so hopefully it won't be an issue.
```
```suggestion
The expectation is that this is a rare pattern and perhaps is not even in use in production, but is worth addressing in the event this applies to your use case.
```

joshuahannan commented:
Thanks for the comments everyone! Lots of great suggestions and edits. I addressed all the comments in the latest commit. @gregsantos I also added a section about what Ledger users need to do to swap.
